Add an (off by default) code path to crash whenever the CoreIPCSecureCoding wrapper is used

beidson opened the pull request:
#### 415a49b8e806d09b279c29b35bf028a3e32fd0dd
<pre>
Add an (off by default) code path to crash whenever the CoreIPCSecureCoding wrapper is used
<a href="https://bugs.webkit.org/show_bug.cgi?id=266033">https://bugs.webkit.org/show_bug.cgi?id=266033</a>
<a href="https://rdar.apple.com/119344064">rdar://119344064</a>

Reviewed by Alex Christensen.

This &quot;crash on encountering an unspecified NSSecureCoding type&quot; mode can be turned on via NSUserDefaults.

* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::applyProcessCreationParameters):
* Source/WebKit/Shared/AuxiliaryProcessCreationParameters.h:
* Source/WebKit/Shared/AuxiliaryProcessCreationParameters.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h:
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
(WebKit::SecureCoding::internalShouldCrashOnSecureCoding):
(WebKit::SecureCoding::internalClassNamesExemptFromSecureCodingCrash):
(WebKit::SecureCoding::shouldCrashOnSecureCoding):
(WebKit::SecureCoding::classNamesExemptFromSecureCodingCrash):
(WebKit::SecureCoding::applyProcessCreationParameters):
(WebKit::crashWithClassName):
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::auxiliaryProcessParameters):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/271746@main">https://commits.webkit.org/271746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/341da47e62a581f770f8992ce69a8b383ec78abf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29496 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8164 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30822 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32042 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26740 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/10373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5456 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26743 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29769 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6819 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25207 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5835 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5962 "Unexpected infrastructure issue, retrying build") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33384 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26925 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26719 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32191 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5925 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4112 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29964 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7652 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/7011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6462 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6448 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
```diff
@@ -43,10 +45,48 @@
+    static dispatch_once_t onceToken;
+    static bool crashOnSecureCoding { false };
+    dispatch_once(&onceToken, ^{
+        NSArray *array = [[NSUserDefaults standardUserDefaults] arrayForKey:@"WebKitCrashOnSecureCodingKey"];
```
Only the UI process should have access to NSUserDefaults.  We're going to have to send a bool in WebProcessCreationParameters, NetworkProcessCreationParameters, GPUProcessCreationParameters, then initialize their process-global value from that bool.

beidson replied:
*Ack*

Yup!

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
```diff
@@ -29,9 +29,70 @@
+        if (!array)
+            return;
+
+        for (id object in array) {
```
May want to make sure it's actually an NSArray before iterating.

beidson replied:
That's guaranteed by this NSUserDefaults API (if it's not an array then nil is returned.)
But I'll put a release assert in.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
```diff
@@ -29,9 +29,70 @@
+    return exemptClassNames.get();
+}
+
+bool shouldCrashOnSecureCoding()
```
...Failure?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
```diff
@@ -29,9 +29,70 @@
+        if (!internalShouldCrashOnSecureCoding())
+            return;
+
+        NSArray *array = [[NSUserDefaults standardUserDefaults] arrayForKey:@"WebKitCrashOnSecureCodingKey"];
```
If the key is to be used for an array of classes to not crash on failure, I think it should have a different name.

achristensen07 commented on Source/WebKit/Shared/AuxiliaryProcessCreationParameters.serialization.in:
```diff
@@ -24,4 +24,6 @@ struct WebKit::AuxiliaryProcessCreationParameters {
     String webCoreLoggingChannels;
     String webKitLoggingChannels;
+    bool crashOnCoreIPCSecureCoding;
+    HashSet<String> classNamesExemptFromSecureCodingCrash;
```
Couldn't these be combined to a std::optional<HashSet<String>>?  It wouldn't make sense to have false then a non-empty set.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
```diff
@@ -29,9 +29,60 @@
+        if (!object)
+            return;
+
+        exemptClassNames.get() = WTF::makeUnique<HashSet<String>>();
```
Using std::optional would prevent an unnecessary allocation, but it would require more memory in the common case where this isn't used.  This is probably better.

beidson committed:
Add an (off by default) code path to crash whenever the CoreIPCSecureCoding wrapper is used
https://bugs.webkit.org/show_bug.cgi?id=266033
rdar://119344064

Reviewed by Alex Christensen.

This "crash on encountering an unspecified NSSecureCoding type" mode can be turned on via NSUserDefaults.

* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::applyProcessCreationParameters):
* Source/WebKit/Shared/AuxiliaryProcessCreationParameters.h:
* Source/WebKit/Shared/AuxiliaryProcessCreationParameters.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h:
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
(WebKit::SecureCoding::internalShouldCrashOnSecureCoding):
(WebKit::SecureCoding::internalClassNamesExemptFromSecureCodingCrash):
(WebKit::SecureCoding::shouldCrashOnSecureCoding):
(WebKit::SecureCoding::classNamesExemptFromSecureCodingCrash):
(WebKit::SecureCoding::applyProcessCreationParameters):
(WebKit::crashWithClassName):
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::auxiliaryProcessParameters):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/271746@main
